plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirCompletionContributorBase.kt

```diff
@@ -110,7 +115,40 @@ internal abstract class FirCompletionContributorBase<C : FirRawPositionCompletio
         }
         priority?.let { lookup.priority = it }
         applyWeighers(context, lookup, symbol, substitutor)
-        sink.addElement(lookup)
+        sink.addElement(lookup.boldIfImmediate())
+    }
+
+    private fun LookupElement.boldIfImmediate(): LookupElement {
```

Comment:
This function and following enum `Style` are very specific to callable completion and should not be in `FirCompletionContributorBase`.  Let's move it (and it application) to the place which is more specific to callable completion.

Reply:
Regarding placement of the styling logic, it would always be needed whenever `addCallableSymbolToCompletion` is invoked, so it seems this is the right place for such logic. Moving it to `FirCallableCompletionContributor` seems inappropriate because this is also needed by `FirSuperMemberCompletionContributor`, which does not inherit `FirCallableCompletionContributor` (and it can't inherit that because they are using different completion context.)

On the other hand, I do not like having this styling logic depending on `CallableWeigher` since it appears they should be totally decoupled from each other. So maybe we should move the logic computing `callableWeight` somewhere else. But then, I don't know where it should go to. Also, doing it that way may make future reader of this code wonder: why isn't the logic computing callable weight not inside the `CallableWeigher`?

In addition, in order to insert type cast, it seems natural to do it also at this place because the computation done in `CallableWeigher` already tells us whether a type cast is needed.  On the other hand, it also seems consistent to implement type cast insertion as a new insertion strategy just like inserting super type qualifier. But that means we need to duplicate the logic in `CallableWeigher` to check if receiver type matches, which can be expensive. What do you think?

Reply:
I just pushed another update with a tentative implementation of inserting type cast on demand. TBH I am not a big fan of the way it's implemented but I can't think of a better way to do it unless we restructure a lot of things. Let me know what you think and I am happy to apply changes to make things cleaner.

Reply:
> I do not like having this styling logic depending on CallableWeigher since it appears they should be totally decoupled from each other
they both may depend on some other part instead :) So, instead of using `CallableWeigher.CallableWeightKind` in styling, some new enum entry may be introduced. And used in weighter as in styling

>  But that means we need to duplicate the logic in CallableWeigher to check if receiver type matches
can we precomute it and put into some context?

Reply:
OK, moved this logic to a helper named `CallableMetadataProvider`.